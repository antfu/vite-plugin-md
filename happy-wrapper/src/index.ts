export { Document, DocumentFragment as Fragment, IElement, IText } from 'happy-dom'

export * from './attributes'
export * from './create'
export * from './diagnostics'
export * from './errors'
export * from './happy-types'
export * from './nodes'
export * from './select'
export * from './type-guards'
export * from './utils'
