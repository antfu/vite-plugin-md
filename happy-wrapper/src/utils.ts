import { flow, pipe } from 'fp-ts/lib/function'
import { createDocument, createElement, createFragment, createTextNode } from './create'
import { HappyMishap } from './errors'
import type { Container, ContainerOrHtml, HTML, NodeSolver, NodeSolverReceiver, NodeType } from './happy-types'
import { isDocument, isElement, isElementLike, isFragment, isTextNode } from './type-guards'

/**
 * Determines the "content-type" of a given node
 */
export const getNodeType = (node: Container | HTML): NodeType => {
  if (typeof node === 'string')
    return 'html'

  return isTextNode(node)
    ? 'text'
    : isElement(node)
      ? 'element'
      : isDocument(node)
        ? 'document'
        : isFragment(node)
          ? 'fragment'
          : 'node'
}

/**
 * A helper utility to help convert DOM nodes or HTML to common type.
 *
 * Start by providing the _exclusions_ you want to make for input. By default, all
 * `Container` types are allowed along with `HTML`
 */
export const solveForNodeType: NodeSolver = (_ = undefined as never) => {
  const solver = <EE extends NodeType, OO>(): NodeSolverReceiver<EE, OO> => ({
    solver: s =>
      (node, parent) => {
        if (node === null)
          throw new Error('Value passed into solver was NULL!')
        if (node === undefined)
          throw new Error('Value passed into solver was UNDEFINED!')

        const type = getNodeType(node)
        if (type in s) {
          const fn = (s as any)[type]
          return fn(node, parent)
        }
        else {
          if (type === 'node' && 'element' in s && isElement(node)) {
            const fn = (s as any).element
            return fn(node, parent)
          }
          else if (type === 'node' && 'text' in s && isTextNode(node)) {
            const fn = (s as any).text
            return fn(node)
          }

          throw new HappyMishap(`Problem finding "${type}" in solver.`, { name: `solveForNodeType(${type})` })
        }
      }
    ,
  })
  return {
    outputType: () => solver(),
    mirror: () => solver(),
  }
}

/**
 * Ensures any Container, array of Containers, or even HTML or HTML[] are all
 * normalized down to just HTML.
 */
export function toHtml<D extends ContainerOrHtml | ContainerOrHtml[] | null>(node: D): HTML {
  if (node === null)
    return ''

  const n = (!Array.isArray(node) ? [node] : node) as ContainerOrHtml[]
  try {
    const results = n.map((i) => {
      const convert = solveForNodeType()
        .outputType<HTML>()
        .solver({
          html: h => h,
          text: t => t.textContent,
          element: e => e.outerHTML,
          node: (ne) => {
            if (isElement(ne))
              convert(ne)
            if (isTextNode(ne))
              convert(ne)

            throw new Error(
              `Unknown node type detected while converting to HTML: [ name: ${ne.nodeName}, type: ${ne.nodeType}, value: ${ne.nodeValue} ]`,
            )
          },
          document: d => `<html>${d.head.hasChildNodes() ? d.head.outerHTML : ''}${d.body.outerHTML}</html>`,
          fragment: (f) => {
            if (isElementLike(f))
              return f.firstElementChild.outerHTML

            else
              return f.childNodes.map(c => convert(c, f)).join('')
          },
        })

      return convert(i)
    })

    return results.join('')
  }
  catch (e) {
    if (Array.isArray(node))
      throw new HappyMishap(`Problem converting an array of ${n.length} nodes [${n.map(i => getNodeType(i as any)).join(', ')}] to HTML`, { name: 'toHTML([...])', inspect: ['first node', node[0]], error: e })

    else
      throw new HappyMishap(`Problem converting "${getNodeType(node)}" to HTML!`, { name: 'toHTML(getNodeType(node))', inspect: node, error: e })
  }
}

/**
 * Clones most DOM types
 */
export function clone<T extends Container | HTML>(container: T): T {
  const clone = solveForNodeType()
    .mirror()
    .solver({
      html: h => `${h}`,
      fragment: flow(toHtml, createFragment),
      document: (d) => {
        return createDocument(d.body.innerHTML, d.head.innerHTML)
      },
      element: e => pipe(e, toHtml, createElement),
      node: flow(toHtml, createFragment, f => f.firstElementChild ? f.firstElementChild : f.firstChild),
      text: flow(toHtml, createTextNode),
    })

  return clone(container)
}

/**
 * ensures that a given string doesn't have any HTML inside of it
 */
export function safeString(str: string): string {
  const node = createFragment(str)
  return node.textContent
}
