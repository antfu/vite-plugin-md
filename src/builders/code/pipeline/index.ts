export * from './addLanguage'
export * from './addClipboard'
export * from './convertBlocksToDomNodes'
export * from './defaultBlocks'
export * from './extractMarkdownItTokens'
export * from './expandCodeBlockVariables'
export * from './inlineStyles'
export * from './highlightLines'
export * from './renderHtml'
export * from './updateLineNumbers'
export * from './useHighlighter'
export * from './userRules'
export * from './updatePreWrapper'
export * from './updateFrontmatterWithCodeBlock'
export * from './updateCodeBlockWrapper'
