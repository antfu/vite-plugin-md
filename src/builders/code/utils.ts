
import type { CodeFenceMeta, CodeOptions, LineCallback, PrismOptions, ShikiOptions } from './types'
import { Highlighter } from './types'

export function isValidLanguage(h: Highlighter, _meta: CodeFenceMeta): boolean {
  if (h === Highlighter.prism) {
    //
  }
  else {
    //
  }
}

/** Type Guard to detect options hash for Prism */
export function usesPrismHighlighting(h: CodeOptions): h is PrismOptions {
  return h.engine === Highlighter.prism
}

/** Type Guard to detect options hash for Shiki */
export function usesShikiHighlighting(h: CodeOptions): h is ShikiOptions {
  return h.engine === Highlighter.shiki
}

export function isLineCallback(cb?: false | string | LineCallback): cb is LineCallback {
  return !!(cb && typeof cb === 'function')
}
