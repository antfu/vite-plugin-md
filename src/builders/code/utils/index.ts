export * from './prism-languages'
export * from './highlighting'
export * from './trace'
export * from './log'
