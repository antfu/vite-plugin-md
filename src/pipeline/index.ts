export * from './applyMarkdownItOptions'
export * from './baseStyling'
export * from './composeSfcBlocks'
export * from './createParser'
export * from './convertHandlerToTask'
export * from './convertToDom'
export * from './escapeCodeTagInterpolation'
export * from './extractBlocks'
export * from './extractFrontmatter'
export * from './finalize'
export * from './pipelineUtilityFunctions'
export * from './frontmatterPreprocess'
export * from './loadMarkdownItPlugins'
export * from './parse'
export * from './repairFrontmatterLinks'
export * from './transformsBefore'
export * from './wrapHtml'
