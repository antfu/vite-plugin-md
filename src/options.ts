import { preprocessHead } from './head'
import { Options, ResolvedOptions } from './types'
import { toArray } from './utils'

export function resolveOptions(userOptions: Options): ResolvedOptions {
  const options = Object.assign({
    headEnabled: false,
    headField: '',
    frontmatter: true,
    exposeFrontmatter: true,
    customSfcBlocks: ['route', 'i18n', 'style'],
    markdownItOptions: {},
    markdownItUses: [],
    markdownItSetup: () => {},
    wrapperClasses: 'markdown-body',
    wrapperComponent: null,
    transforms: {},
    frontmatterPreprocess: (frontmatter: any, options: ResolvedOptions) => {
      const head = preprocessHead(frontmatter, options)
      return { head, frontmatter }
    },
  }, userOptions) as ResolvedOptions

  options.wrapperClasses = toArray(options.wrapperClasses).filter(i => i).join(' ')

  return options
}
