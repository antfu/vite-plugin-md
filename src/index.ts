import type { Plugin } from 'vite'
import { createFilter } from '@rollup/pluginutils'
import { createMarkdown } from './markdown'
import { resolveOptions } from './options'
import type { Options } from './types'

function VitePluginMarkdown(userOptions: Options = {}): Plugin {
  const options = resolveOptions(userOptions)
  const markdownToVue = createMarkdown(options)

  const filter = createFilter(
    userOptions.include || /\.md$/,
    userOptions.exclude,
  )

  return {
    name: 'vite-plugin-md',
    enforce: 'pre',
    transform(raw, id) {
      if (!filter(id))
        return
      try {
        return markdownToVue(id, raw)
      }
      catch (e: any) {
        this.error(e)
      }
    },
    async handleHotUpdate(ctx) {
      if (!filter(ctx.file))
        return

      const defaultRead = ctx.read
      ctx.read = async function() {
        return markdownToVue(ctx.file, await defaultRead())
      }
    },
  }
}

export default VitePluginMarkdown
