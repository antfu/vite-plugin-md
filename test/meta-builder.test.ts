import { readFile } from 'fs/promises'
import { beforeAll, describe, expect, it } from 'vitest'
import { resolveOptions } from '../src/options'
import { meta } from '../src/index'
import type { Options } from '../src/types'
import { composeSfcBlocks } from '../src/pipeline'
import { composeFixture } from './utils'

let md = ''

describe('use "meta" builder for frontmatterPreprocess', async () => {
  beforeAll(async () => {
    md = await readFile('test/fixtures/meta.md', 'utf-8')
  })

  it('with no config, doc props all available as frontmatter props and other meta props get default mapping', async () => {
    const sfc = await composeSfcBlocks('', md, { builders: [meta()] })

    expect(sfc.frontmatter.title).toEqual('Metadata Rules')
    expect(sfc.frontmatter.byline).toEqual('who loves ya baby?')
    expect(sfc.frontmatter.layout).toEqual('yowza')
    expect(sfc.frontmatter.image).toEqual('facebook.png')

    expect(sfc.head.title).toEqual('Metadata Rules')
    expect(sfc.routeMeta?.meta?.layout).toEqual('yowza')
    expect(sfc.meta.find(p => p.key === 'title')).toBeDefined()
    expect(sfc.meta.find(p => p.key === 'image')).toBeDefined()
  })

  it('default value is used when no frontmatter is present', async () => {
    const options: Options = {
      frontmatterDefaults: {
        title: 'nada',
        description: 'there I was, there I was',
      },
      builders: [
        meta(),
      ],
    }
    const sfc = await composeSfcBlocks('', md, resolveOptions(options))

    expect(
      sfc.frontmatter.title,
      'default value should have been ignored in favor of page value',
    ).toBe('Metadata Rules')

    expect(
      sfc.frontmatter.description,
      `default value should have presented, found: ${sfc.frontmatter}`,
    ).toBe('there I was, there I was')

    expect(
      sfc.meta.find(i => i.key === 'description'),
      'description, as a default value, should now be in meta props',
    ).toBeDefined()
  })

  it('frontmatter props exported', async () => {
    const { component } = await composeFixture('meta')

    expect(component.includes('const title')).toBeTruthy()
    expect(component.includes('const byline')).toBeTruthy()
    expect(component.includes('const layout')).toBeTruthy()
    expect(component.includes('const image')).toBeTruthy()

    expect(component.includes('const frontmatter')).toBeTruthy()
    expect(component.includes('export const frontmatter')).toBeTruthy()
    expect(component.includes('defineExpose({ ')).toBeTruthy()
  })
})

describe('meta() can manage route meta', () => {
  it('router not brought into script section when Markdown doesn\'t have route meta', async () => {
    const { scriptSetup } = await composeFixture('simple', { builders: [meta()] })

    expect(scriptSetup).not.toContain('useRouter')
  })

  it('router IS imported when a a \'route prop\' is defined in frontmatter', async () => {
    const { scriptSetup, frontmatter } = await composeFixture('meta', { builders: [meta()] })
    expect(frontmatter.layout).toBeDefined()
    expect(scriptSetup).toContain('useRouter')
  })

  it('using routeName callback ensures that router is imported and route name is set', async () => {
    const { frontmatter, scriptSetup } = await composeFixture('no-route', {
      builders: [meta({
        routeName: (filename, fm) => fm.title ? `bespoke-${fm.title}` : 'nada',
      })],
    })

    expect(frontmatter.title).toBe('NoRoute')
    expect(scriptSetup, 'should have included an import of useRouter!').toContain('useRouter')
    expect(
      scriptSetup,
      `The scriptSetup block was:\n${scriptSetup}\n\n`,
    ).toContain('router.currentRoute.value.name = "bespoke-NoRoute"')
  })
})

describe('meta() snapshots', async () => {
  beforeAll(async () => {
    md = await readFile('test/fixtures/meta.md', 'utf-8')
  })

  it('frontmatter is consistent', async () => {
    const { frontmatter } = await composeSfcBlocks('/foobar/meta.md', md)
    expect(frontmatter).toMatchSnapshot()
  })

  it('HTML is consistent', async () => {
    const { html } = await composeSfcBlocks('/foobar/meta.md', md)
    expect(html).toMatchSnapshot()
  })

  it('script blocks are consistent', async () => {
    const { scriptBlocks } = await composeSfcBlocks('/foobar/meta.md', md)
    expect(scriptBlocks).toMatchSnapshot()
  })

  it('custom blocks are consistent', async () => {
    const { customBlocks } = await composeSfcBlocks('/foobar/meta.md', md)
    expect(customBlocks).toMatchSnapshot()
  })
})
