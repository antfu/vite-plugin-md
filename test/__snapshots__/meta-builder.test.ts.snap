// Vitest Snapshot v1

exports[`meta() snapshots > HTML is consistent 1`] = `
"<div class=\\"markdown-body\\"><h1>Metadata</h1>
<blockquote>
<p>{{ byline }}</p>
</blockquote>
<p>This is a markdown file intended for testing meta data.</p>
<p><img src=\\"{{image}}\\" alt=\\"image\\">
<img src=\\"{{ image }}\\"></p>
</div>"
`;

exports[`meta() snapshots > custom blocks are consistent 1`] = `[]`;

exports[`meta() snapshots > frontmatter is consistent 1`] = `
{
  "byline": "who loves ya baby?",
  "image": "facebook.png",
  "layout": "yowza",
  "name": "MetaBuilder",
  "title": "Metadata Rules",
}
`;

exports[`meta() snapshots > script blocks are consistent 1`] = `
[
  "
<script lang=\\"ts\\">
/** frontmatter meta-data for MD page **/
export interface Frontmatter {
  title?: string; description?: string; subject?: string; category?: string; name?: string; excerpt?: string; image?: string; layout?: string; requiredAuth?: boolean; meta?: Record<string, any>[];
  [key: string]: unknown
}
export const frontmatter: Frontmatter = {\\"title\\":\\"Metadata Rules\\",\\"name\\":\\"MetaBuilder\\",\\"byline\\":\\"who loves ya baby?\\",\\"layout\\":\\"yowza\\",\\"image\\":\\"facebook.png\\"}
export const excerpt: string | undefined = \\"\\"

</script>",
]
`;
